blog: Add post on introducing Kubeflow Trainer V2 (#169)
kubeflow/blog: _posts/2025-07-09-introducing-trainer-v2.md

---
toc: true
layout: post
categories: [ trainer ]
comments: true
title: "Introducing Kubeflow Trainer V2"
hide: false
permalink: /trainer/intro/
author: "AutoML & Training WG"
---

Running machine learning workloads on Kubernetes can be challenging. Distributed training, in particular, involves managing multiple nodes, GPUs, large datasets, and fault tolerance, which often requires deep Kubernetes knowledge. The **Kubeflow Trainer v2 (KF Trainer)** was created to simplify this complexity, by making training on Kubernetes easier for data scientists and machine learning engineers.

**The main goals of KF Trainer v2 include:**
- Make AI/ML workloads easier to manage at scale
- Improve the Python interface
- Add built-in support for fine-tuning large language models

> - Abstract Kubernetes complexity from data scientists

Reviewer: ```suggestion
- Abstract Kubernetes complexity from model developers
```

Author: I think that makes sense - ML engineers / data scientists is already used on the Kubeflow website
What do you think @franciscojavierarceo ?

Author: Having that we've agreed on https://github.com/kubeflow/website/pull/4144, shall we just use AI Practitioners here?
```suggestion
- Abstract Kubernetes complexity from AI Practitioners
```
Any thoughts @andreyvelich @franciscojavierarceo ?

Reviewer 2: Yeah, as we discussed here, let's call them AI Practitioners: https://github.com/kubeflow/blog/pull/169#discussion_r2208998134

**Kubeflow Trainer v2** represents the next evolution of the **Kubeflow Training Operator**, building on over seven years of experience running ML workloads on Kubernetes. The journey began in 2017 when the **Kubeflow** project introduced **TFJob** to orchestrate TensorFlow training on Kubernetes. At that time, Kubernetes lacked many of the advanced batch processing features needed for distributed ML training, so the community had to implement these capabilities from scratch.

Over the years, the project expanded to support multiple ML frameworks including **PyTorch**, **MXNet**, **MPI**, and **XGBoost** through various specialized operators. In 2021, these were consolidated into the unified **Training Operator v1**. Meanwhile, the Kubernetes community introduced the **Batch Working Group**, developing important APIs like JobSet, Kueue, Indexed Jobs, and PodFailurePolicy that improved HPC and AI workload management.

**Trainer v2** leverages these Kubernetes-native improvements to not reimplement batch orchestration features. This collaboration between the Kubernetes and Kubeflow communities delivers a more standardized approach to ML training on Kubernetes.

# Division of Labor

> One of the main challenges with ML training on Kubernetes is that it often requires **data scientists** to have an understanding of **Kubernetes concepts** and the **infrastructure** being used for training. This distracts data scientists from their primary focus.

Reviewer 2: FYI, we name them ML Users in the Trainer docs right now: https://www.kubeflow.org/docs/components/trainer/overview/#user-personas

> One of the main challenges with ML training on Kubernetes is that it often requires **data scientists** to have an understanding of **Kubernetes concepts** and the **infrastructure** being used for training. This distracts data scientists from their primary focus.

Reviewer 2: As we discussed here, we will use AI Practitioners: https://github.com/kubeflow/website/pull/4144

# Python SDK

**The KF Trainer v2** introduces a **redesigned Python SDK**, which is intended to be the **primary interface for data scientist users**. The SDK provides a unified interface across multiple ML frameworks and cloud environments, abstracting away the underlying Kubernetes complexity.

The diagram below illustrates how Kubeflow Trainer provides a unified layer across different ML frameworks, Kubernetes infrastructure and cloud providers:

![trainerv2](/images/2025-07-09-introducing-trainer-v2/trainerv2.png)

**Kubeflow Trainer v2** supports multiple ML frameworks through **pre-configured runtimes**. The table below shows the current framework support:

![runtimes](/images/2025-07-09-introducing-trainer-v2/runtimes.png)

The SDK makes it easier for users familiar with Python to **create, manage, and monitor training jobs**, without requiring them to deal with any YAML definitions:

```
from kubeflow.trainer import TrainerClient

client = TrainerClient()

job_name = client.train(
  runtime=client.get_runtime("torch-distributed"),
  trainer=CustomTrainer(

>     func=my_train_func,

Reviewer 2: @kramaranya Did you get a chance to check it ?

# Extensibility and Pipeline Framework

One of the challenges in **KF Trainer v1** was supporting additional ML frameworks, especially for closed-sourced frameworks. The v2 architecture addresses this by introducing a **Pipeline Framewor**k that allows customers to **extend the Plugins** and **support orchestration** for their custom in-house ML frameworks.

The diagram below shows Kubeflow Trainer Pipeline Framework overview:

![trainer_pipeline_framework](/images/2025-07-09-introducing-trainer-v2/trainer_pipeline_framework.drawio.svg)

The framework works through a series of phases - **Startup**, **PreExecution**, **Build**, and **PostExecution** - each with **extension points** where custom Plugins can hook in. This approach allows adding support for new frameworks, custom validation logic, or specialized training orchestration without changing the underlying system.

# LLM Fine-Tuning Support

Another improvement of **Trainer v2** is its **built-in support for fine-tuning large language models**, where we provide two types of trainers:

> - `BuiltinTrainer` - already includes the fine-tuning logic and allows data scientists to quickly start fine-tuning requiring only parameter adjustments,

Author: thank you! updated in https://github.com/kubeflow/blog/pull/169/commits/336b058398f8cbcc23e0f50a7e440727545b486d

---
toc: true
layout: post
categories: [ trainer ]
comments: true
title: "Introducing Kubeflow Trainer V2"
hide: false
permalink: /trainer/intro/

> author: "AutoML & Training WG"

Reviewer 3: Maybe "Kubeflow Trainer Team"?

# What’s Next?

Future enhancements will continue to improve the user experience, integrate deeper with other Kubeflow components, and support more training frameworks. **Upcoming features** include:
- **[Local Execution](https://github.com/kubeflow/sdk/issues/22)** - run training jobs locally without Kubernetes
- **[Unified Kubeflow SDK](https://docs.google.com/document/d/1rX7ELAHRb_lvh0Y7BK1HBYAbA0zi9enB0F_358ZC58w/edit?tab=t.0#heading=h.e0573r7wwkgl)** - a single SDK for all Kubeflow components 
- **[Trainer UI](https://github.com/kubeflow/trainer/issues/2648)** - a user interface to expose high level metrics for training jobs and monitor training logs
- **[Native Kueue integration](https://github.com/kubernetes-sigs/kueue/issues/3884)** - improve resource management and scheduling capabilities for TrainJob resources
- **[Model Registry integrations](https://github.com/kubeflow/trainer/issues/2245)** - export trained models directly to Model Registry

For users migrating from **Trainer v1**, check out a [**Migration Guide**](https://www.kubeflow.org/docs/components/trainer/operator-guides/migration/).

> # Resources and Community

Author: I already mention them https://github.com/kramaranya/blog/blame/trainer-v2/_posts/2025-07-09-introducing-trainer-v2.md#L23-L25, do you want to duplicate it?

---
toc: true
layout: post
categories: [ trainer ]
comments: true
title: "Introducing Kubeflow Trainer V2"
hide: false
permalink: /trainer/intro/
author: "AutoML & Training WG"
---

> Running machine learning workloads on Kubernetes can be challenging. Distributed training, in particular, involves managing multiple nodes, GPUs, large datasets, and fault tolerance, which often requires deep Kubernetes knowledge. The **Kubeflow Trainer v2 (KF Trainer)** was created to simplify this complexity, by making training on Kubernetes easier for AI Practitioners.

Reviewer 4: nit: "hide this complexity"

> **The main goals of KF Trainer v2 include:**

Reviewer 4: ```suggestion
**The main goals of Kubeflow Trainer v2 include:**
```

> **The main goals of KF Trainer v2 include:**

Reviewer 4: It may be obvious for everyone involved in the project but it doesn't seem to me like very explicit / prominent in this article: PyTorch :) I'd try to message that Kubeflow trainer v2 is the easiest and more scalable way to run PyTorch distributed training on Kubernetes!

Author: Yeah, I do agree we should emphasize on this point.

I'm leaning toward modifying the current goal "Make AI/ML workloads easier to manage at scale" to be:
"**Make AI/ML workloads easier to manage at scale, with PyTorch as the primary framework**"

And then modify an intro:
"Running machine learning workloads on Kubernetes can be challenging. Distributed training and LLMs fine-tuning, in particular, involves managing multiple nodes, GPUs, large datasets, and fault tolerance, which often requires deep Kubernetes knowledge. The Kubeflow Trainer v2 (KF Trainer) was created to hide this complexity, by abstracting Kubernetes from AI Practitioners and **providing the easiest, most scalable way to run distributed PyTorch jobs.**"

Alternately, we could just add a new goal with no intro chnages:
"Deliver the easiest and most scalable PyTorch distributed training on Kubernetes"

What do you think @astefanutti @andreyvelich ?

Author: @andreyvelich could you also take a look at ^^, so I can update it?

Author: Awesome, updated in https://github.com/kubeflow/blog/pull/169/commits/aa4f6e745824b7eb110f786d312139b9b8c9c81d. @astefanutti please let me know what you think :)

Reviewer 4: This looks great, thanks!

- Make AI/ML workloads easier to manage at scale

> - Improve the Python interface

Reviewer 4: ```suggestion
- Provide a Pythonic interface to train models
```

- Add built-in support for fine-tuning large language models
- Abstract Kubernetes complexity from AI Practitioners
- Consolidate efforts between Kubernetes Batch WG and Kubeflow community

> We’re deeply grateful to all contributors and community members who made the **Trainer v2** possible with their hard work and valuable feedback. We'd like to give special recognition to [andreyvelich](https://github.com/andreyvelich), [tenzen-y](https://github.com/tenzen-y), [electronic-waste](https://github.com/electronic-waste), [astefanutti](https://github.com/astefanutti), [ironicbo](https://github.com/ironicbo), [mahdikhashan](https://github.com/mahdikhashan), [kramaranya](https://github.com/kramaranya), [harshal292004](https://github.com/harshal292004), [akshaychitneni](https://github.com/akshaychitneni), [chenyi015](https://github.com/chenyi015) and the rest of the contributors. We would also like to highlight [ahg-g](https://github.com/ahg-g), [kannon92](https://github.com/kannon92), and [vsoch](https://github.com/vsoch) whose feedback was essential while we designed the Kubeflow Trainer architecture together with the Batch WG. See the full [contributor list](https://kubeflow.devstats.cncf.io/d/66/developer-activity-counts-by-companies?orgId=1&var-period_name=Last%206%20months&var-metric=commits&var-repogroup_name=kubeflow%2Ftrainer&var-country_name=All&var-companies=All) for everyone who helped make this release possible.

Reviewer 4: nit: break lines to keep one sentence per line.

Over the years, the project expanded to support multiple ML frameworks including **PyTorch**, **MXNet**, **MPI**, and **XGBoost** through various specialized operators. In 2021, these were consolidated into the unified **[Training Operator v1](https://docs.google.com/document/d/1x1JPDQfDMIbnoQRftDH1IzGU0qvHGSU4W6Jl4rJLPhI/edit?tab=t.0#heading=h.e33ufidnl8z6)**. Meanwhile, the Kubernetes community introduced the **Batch Working Group**, developing important APIs like JobSet, Kueue, Indexed Jobs, and PodFailurePolicy that improved HPC and AI workload management.

**Trainer v2** leverages these Kubernetes-native improvements to re-use existing functionality and not reinventing the wheel. This collaboration between the Kubernetes and Kubeflow communities delivers a more standardized approach to ML training on Kubernetes.

> # Division of Labor

Reviewer 4: Labor sounds a bit too laborious 😃. Maybe just "User Personas" or "For AI practitioners and MLOps engineers"?

Reviewer 2: @kramaranya Did you get a chance to check it ?

Reviewer 2: Sure, User Personas make sense to me.

The SDK provides a unified interface across multiple ML frameworks and cloud environments, abstracting away the underlying Kubernetes complexity.

The diagram below illustrates how Kubeflow Trainer provides a unified layer across different ML frameworks, Kubernetes infrastructure and cloud providers:

![trainerv2](/images/2025-07-09-introducing-trainer-v2/trainerv2.png)

**Kubeflow Trainer v2** supports multiple ML frameworks through **pre-configured runtimes**. The table below shows the current framework support:

![runtimes](/images/2025-07-09-introducing-trainer-v2/runtimes.png)

The SDK makes it easier for users familiar with Python to **create, manage, and monitor training jobs**, without requiring them to deal with any YAML definitions:

```
from kubeflow.trainer import TrainerClient

client = TrainerClient()

def my_train_func():
    """User defined function that runs on each distributed node process"""
    import os
    import torch
    import torch.distributed as dist
    from torch.utils.data import DataLoader, DistributedSampler
    
    # Setup PyTorch distributed
    local_rank = int(os.getenv("LOCAL_RANK", 0))

>     dist.init_process_group(backend=backend)

Reviewer 2: If you want to add these 2 lines, let's also add this:
```
backend = "nccl" if torch.cuda.is_available() else "gloo"
```

This integration allows the KF Trainer v2 to better utilize standard Kubernetes features instead of trying to recreate them.

# Kueue Integration

Resource management is improved through integration with **[Kueue](https://kueue.sigs.k8s.io/)**, a **Kubernetes-native queueing system**.
The KF Trainer v2 includes initial support for Kueue through Pod Integration, which allows individual training pods to be queued when resources are busy.
We are working on **[native Kueue support](https://github.com/kubernetes-sigs/kueue/issues/3884)** for `TrainJob` to provide richer queueing features in future releases.

# MPI Support

The **KF Trainer v2** also provides **MPI support**, which includes **automatic generation of SSH keys** for secure inter-node communication.

![MPI_support](/images/2025-07-09-introducing-trainer-v2/MPI-support.drawio.svg)

The diagram above shows how this works in practice - the **KF Trainer** automatically **handles the SSH key generation** and **MPI communication** between training pods, which allows frameworks like DeepSpeed to coordinate training across multiple GPU nodes without requiring manual configuration of inter-node communication.

> # Fault Tolerance Improvements

Author: I've added Gang-Scheduling section in https://github.com/kubeflow/blog/pull/169/commits/f2cfa0f853422026a737425f6beed40414cffc5d. Please review it and let me know if any changes needed.

# User Personas

One of the main challenges with ML training on Kubernetes is that it often requires **AI Practitioners** to have an understanding of **Kubernetes concepts** and the **infrastructure** being used for training. This distracts AI Practitioners from their primary focus.

**The KF Trainer v2** addresses this by **separating the infrastructure configuration from the training job definition**.
This separation is built around three new custom resources definitions (CRDs):
- `TrainingRuntime` - a namespace-scoped resource that contains the infrastructure details that are required for a training job, such as the training image to use, failure policy, and gang-scheduling configuration.
- `ClusterTrainingRuntime` - similar to `TrainingRuntime`, but cluster scoped.
- `TrainJob` - specifies the training job configuration, including the training code to run, config for pulling the training dataset & model, and a reference to the training runtime.

The diagram below shows how different personas interact with these custom resources:

![user_personas](/images/2025-07-09-introducing-trainer-v2/user-personas.drawio.svg)

> - **Platform Engineers** define and manage **the infrastructure configurations** required for training jobs using `TrainingRuntimes` or `ClusterTrainingRuntimes`. 

Author: Hm, @andreyvelich should it actually be Platform Administrators?

Reviewer 2: Yes, let's keep the persona name consistent please.

Author: thanks, updated in https://github.com/kubeflow/blog/pull/169/commits/4280688ae9da46460d618f59a3586ad095ef35c5

---
toc: true
layout: post
categories: [ trainer ]
comments: true
title: "Democratizing AI Model Training on Kubernetes: Introducing Kubeflow Trainer V2"
hide: false
permalink: /trainer/intro/
author: "Kubeflow Trainer Team"
---

Running machine learning workloads on Kubernetes can be challenging.
Distributed training and LLMs fine-tuning, in particular, involves managing multiple nodes, GPUs, large datasets, and fault tolerance, which often requires deep Kubernetes knowledge.

> The **Kubeflow Trainer v2 (KF Trainer)** was created to hide this complexity, by abstracting Kubernetes from AI Practitioners and providing the easiest, most scalable way to run distributed PyTorch jobs.

Author: The PyTorch focus is intentional here because we want to emphasize that the main goal of Trainer v2 is specifically to make distributed PyTorch jobs easier to run, see https://github.com/kubeflow/blog/pull/169#discussion_r2215215754

- `TrainingRuntime` - a namespace-scoped resource that contains the infrastructure details that are required for a training job, such as the training image to use, failure policy, and gang-scheduling configuration.
- `ClusterTrainingRuntime` - similar to `TrainingRuntime`, but cluster scoped.
- `TrainJob` - specifies the training job configuration, including the training code to run, config for pulling the training dataset & model, and a reference to the training runtime.

The diagram below shows how different personas interact with these custom resources:

![user_personas](/images/2025-07-09-introducing-trainer-v2/user-personas.drawio.svg)

- **Platform Engineers** define and manage **the infrastructure configurations** required for training jobs using `TrainingRuntimes` or `ClusterTrainingRuntimes`. 
- **AI Practitioners** focus on model development using the simplified `TrainJob` resource or **Python SDK** wrapper, providing a reference to **the training runtime** created by **Platform Engineers**.

# Python SDK

**The KF Trainer v2** introduces a **redesigned Python SDK**, which is intended to be the **primary interface for AI Practitioners**.

> The SDK provides a unified interface across multiple ML frameworks and cloud environments, abstracting away the underlying Kubernetes complexity.

Author: This means you can use the same SDK commands and configurations for any cloud provider, without needing to learn different APIs for each platform. I think 'a unified interface' works better here, comparing to 'the same interface'.
https://www.kubeflow.org/docs/components/trainer/overview/#what-is-kubeflow-trainer

> The diagram below illustrates how Kubeflow Trainer provides a unified layer across different ML frameworks, Kubernetes infrastructure and cloud providers:

Reviewer 5: Suggestion to disambiguate "unified layer"

```suggestion
The diagram below illustrates how Kubeflow Trainer provides a consistent experience for running ML jobs across various different ML frameworks, Kubernetes infrastructures, and cloud providers:
```

    dataset = YourDataset()
    train_loader = DataLoader(dataset, sampler=DistributedSampler(dataset))
    
    # Your training logic here
    for epoch in range(num_epochs):
        for batch in train_loader:
            # Forward pass, backward pass, optimizer step
            ...
            
    # Wait for the distributed training to complete
    dist.barrier()
    if dist.get_rank() == 0:
        print("Training is finished")

    # Clean up PyTorch distributed
    dist.destroy_process_group()

job_name = client.train(
  runtime=client.get_runtime("torch-distributed"),
  trainer=CustomTrainer(
    func=my_train_func,
    num_nodes=5,
    resources_per_node={
      "gpu": 2,
     },
  ),
)

job = client.get_job(name=job_name)

for step in job.steps:
   print(f"Step: {step.name}, Status: {step.status}")

client.get_job_logs(job_name, follow=True)
```
The SDK handles all Kubernetes API interactions. This eliminates the need for AI Practitioners to directly interact with the Kubernetes API.

# Simplified API

Previously, in the **Kubeflow Training Operator** users worked with different custom resources for each ML framework, each with their own framework-specific configurations.

> The **KF Trainer v2** replaces these multiple CRDs with a **unified TrainJob API** that works with **multiple ML frameworks**.

Author: To stay consistent, we should keep `KF Trainer v2`, and I would keep `API` to avoid duplication :)

              imagePullPolicy: Always
              command:
                - "python3"
                - "/opt/pytorch-mnist/mnist.py"
                - "--epochs=1"
```

In the **KF Trainer v2**, creating an equivalent job becomes much simpler:

```
apiVersion: trainer.kubeflow.org/v1alpha1
kind: TrainJob
metadata:
  name: pytorch-simple
  namespace: kubeflow
spec:
  trainer:
    numNodes: 2
    image: docker.io/kubeflowkatib/pytorch-mnist:v1beta1-45c5727
    command:
      - "python3"
      - "/opt/pytorch-mnist/mnist.py"
      - "--epochs=1"
  runtimeRef:
    name: torch-distributed
    apiGroup: trainer.kubeflow.org
    kind: ClusterTrainingRuntime
```

Additional **infrastructure** and **Kubernetes-specific** details are provided in the referenced **runtime** definition, and managed separately by **Platform Engineers**.
In the future, we might support other runtimes in addition to `TrainingRuntime` and `ClusterTrainingRuntime`, for example [SlurmRuntime](https://github.com/kubeflow/trainer/issues/2249).

# Extensibility and Pipeline Framework

One of the challenges in **KF Trainer v1** was supporting additional ML frameworks, especially for closed-sourced frameworks.

> The v2 architecture addresses this by introducing a **Pipeline Framewor**k that allows customers to **extend the Plugins** and **support orchestration** for their custom in-house ML frameworks.

Reviewer 2: Yes, the doc is work in progress by @IRONICBo is here: https://github.com/kubeflow/website/pull/4039

@kramaranya @eoinfennessy Maybe we could be more explicit here, and say that allows platform administrators to **extend the Plugins** ... ?

> The v2 architecture addresses this by introducing a **Pipeline Framewor**k that allows customers to **extend the Plugins** and **support orchestration** for their custom in-house ML frameworks.

Author: updated in https://github.com/kubeflow/blog/pull/169/commits/4280688ae9da46460d618f59a3586ad095ef35c5
